feat(router) ✨ Adds Asynchrounous Redirects

MeddahAbdellah opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/contributing-docs/commit-message-guidelines.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.dev application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
Currently the angular router only allows a string or a synchronous function for the redirectTo field of the router.


## What is the new behavior?
Currently the angular router now allows also asynchronous functions that return a Promise or an Observable.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

I had to change one of the tests whose assertion was unreachable for it to fail. There might be a bug in the redirection strategy. I don't know if it's done on purpose, but if the path that has the redirectTo is a subpath of the redirected to url. The redirectTo function returns it as a UrlTree, which causes an infinite loop. I have been able to reproduce it here:
https://stackblitz.com/edit/angular-vakejq-tmlfa9gy?file=src%2Fapp.routes.ts


I have added a test to lock this behaviour, but I suspect it should be modified, at least throwing an error in devMode telling the user that there might be an infinite loop. I'd gladly do it in another PR once a decision on what should be done is made. In the meantime, this PR adds async management of redirection, keeping all the current behaviours of the simple functional redirection. All tests pass locally.

thePunderWoman commented:
@MeddahAbdellah Thanks for the contribution! Looks like you have some failing tests and lint issues to fix.

MeddahAbdellah committed:
feat(router): add asynchronous redirects

Adds support for asynchronous redirects in the router, allowing redirect logic to be resolved dynamically (e.g., via API or async function). This enhances routing flexibility and supports more complex navigation scenarios.

BREAKING CHANGE: The `RedirectFn` can now return `Observable` or
`Promise`. Any code that directly calls functions returning this type
may need to be adjusted to account for this.

MeddahAbdellah commented:
Apologies, forgot to put the PR in draft. I have updated the typo that caused the tests to fail, I'll launch them again and put the PR to review once everything is green

wartab commented on packages/router/src/apply_redirects.ts:
```diff
@@ -88,31 +88,47 @@ export class ApplyRedirects {
-      }
 
-      redirectTo = newRedirect;
+      if (newRedirect instanceof Observable) {
```
The router actually already has a function that does this if-else chain.
https://github.com/angular/angular/blob/c0ef923ebb409b8554059d13afbea45b38fed952/packages/router/src/utils/collection.ts#L69-L82

atscott commented on goldens/public-api/router/index.api.md:
```diff
@@ -606,7 +606,7 @@ export class RedirectCommand {
 
 // @public
-export type RedirectFunction = (redirectData: Pick<ActivatedRouteSnapshot, 'routeConfig' | 'url' | 'params' | 'queryParams' | 'fragment' | 'data' | 'outlet' | 'title'>) => string | UrlTree;
+export type RedirectFunction = (redirectData: Pick<ActivatedRouteSnapshot, 'routeConfig' | 'url' | 'params' | 'queryParams' | 'fragment' | 'data' | 'outlet' | 'title'>) => string | UrlTree | Observable<string | UrlTree> | Promise<string | UrlTree>;
```
The router also has a common type for the T|Observable<T>|Promise<T> union: https://angular.dev/api/router/MaybeAsync#

MeddahAbdellah replied:
Thank you for the review. I have taken it into account. I couldn't work on the PR during the week, I will finish the tests tonight and put it up for review.

atscott reviewed (approved):
reviewed-for: public-api

thePunderWoman reviewed (approved):
reviewed-for: public-api

atscott commented:
Note to self: This should likely be marked as a breaking change

MeddahAbdellah commented on packages/router/src/apply_redirects.ts:
```diff
@@ -199,3 +196,20 @@ export class ApplyRedirects {
   }
 }
+
+function getRedirectResult(
+  redirectTo: string | RedirectFunction,
+  currentSnapshot: ActivatedRouteSnapshot,
+  injector: Injector,
+): Observable<string | UrlTree> {
+  if (typeof redirectTo === 'string') {
+    return of(redirectTo);
+  }
+  const redirectToFn = redirectTo;
+  const {queryParams, fragment, routeConfig, url, outlet, params, data, title} = currentSnapshot;
+  return wrapIntoObservable(
+    runInInjectionContext(injector, () =>
+      redirectToFn({params, data, queryParams, fragment, routeConfig, url, outlet, title}),
+    ),
+  );
```
Thought (can be ignored): I don't know if it's better to pull it into a new function. I have had a hard time going up the stack trace when trying to debug. granted this is more elegant, I have been finding myself avoiding too much depth in my function calls (I mean by that, a function that calls a function that calls a function), it's easier to debug I find but that remains a personal opinion.
